proposed/2022/Multiple-Equivalent-Framework-Support-TFM-As-Aliases.md

```diff
@@ -0,0 +1,220 @@
+It is not uncommon that people use `nuget.exe restore` due to the fact that they have non-SDK projects in their solution.
+
+Given that NuGet.exe and .NET SDK do not ship together, it is super common that the assets file was generated by a newer/older version of NuGet.exe than the .NET SDK can support. 
+
+Normally, this works. NuGet rarely makes non-additive changes to the assets file.
+
+Sample part of an assets file:
+
+```json
+  "targets": {
+    ".NETFramework,Version=v4.7.2": {},
+    ".NETFramework,Version=v4.8.0": {}
+  },
+
+```
+
+NuGet's assets file is written pivoting on the actual target framework.
+
+To solve this, NuGet would need to:
+
+- Add vNext for the assets file format, version 4. Version 4 would either a dictionary based on the `TargetFramework` value instead of the effective target framework, built by coming `TargetFrameworkMoniker` and `TargetPlatformMoniker`.
+  - This behavior would probably have to be opt-in to avoid breaking all combinations of new `NuGet.exe`, old `.NET SDK` combinations, but we can choose to intentionally not do this to encourage maintaining the 2 versions equivalently.
+- Make a decision what would be a satisfactory behavior for combining newer/older versions of `NuGet.exe` and `.NET SDK`. This is largely just confirming some of the recent decisions.
+
+To solve this, the .NET SDK would need:
+
+- Be able to `build` with both versions 3 and version 4 of the assets file. The NuGet dependencies would naturally flow into the .NET SDK and the work is to just differentiate between the 2 different versions.
+  - The amount of work needed to be done here would likely be affected by the exact shape of the NuGet APIs. These changes may be seamless.
+
+### Restore challenges
+
+In addition to the assets file, restore has a lock file which has a similar schema, that schema would need to be amended. Fortunately when Central Package Management and its transitive pinning was introduced we introduced the concept of a PackagesLockFile version succesfully. We'd just add a version 3.
+
+NuGet would:
+
+- Add version 3 of the packages lock file.
+
+NuGet's intra restore caching is based on the existing of a single framework. It is difficult to predict the amount of work that'd be required to implement this correctly, but it'll certainly involve public API changes to NuGet libraries. (NuGet ships libraries, NuGet.exe, VS and dotnet.exe)
+
+### Pack challenges
+
+Multi-targetted pack only really works when each build leg (TargetFramework value) has 1 target location such as `lib/net5.0`.
+Packing projects that do target the same framework multiple may be helpful in some scenarios, so completely blocking it is likely not a great idea, but having a coherent experience would require better understanding of the scenarios at question.
+
+At the minimum, the pack command should be able to detect the scenario at hand and *not succeed queitly* as there's no guarantee what gets packed.

```

Comment:
One solution on top of my head is mapping folders as per the aliases from the user-defined root (`\tools\vs17.2` or `\lib\banana`) or just user-defined mapping (`\lib\net5.0\Design\vs17.2`).

```xml
<PackageFolderForTarget Include="banana" Folder="lib"/>
<PackageFolderForTarget Include="vs" Folder="tools"/>
<PackageFolderForTarget Include="vs17" Folder="lib" RelativePath="Design\<alias>"/>
<PackageFolderForTarget Include="vs17.3" Folder="lib" RelativePath="Design" FileName="<original>.<alias>.<ext>"/>
```

The last two seems un-imaginative but you get the point, right? You can't use `TargetFramework` or `Identity` here because it could include version-less and limited versioned targets too. In the example above, `vs` could match all `vsXX.X` and is overridden by the following two with similar rules.

Reply:
I think we need to differentiate between something being part of the NuGet package contract and being part of the build contract. 

I'd imagine any solution for would likely need some way to specify the target folders, the items etc as you suggested. Not sure what they looks like just yet, but as we iterate on this we can update.

Reply:
When I update the spec, I'm going to make pack, where the resolved `TargetFrameworkMoniker` and `TargetPlatformMoniker` is not unique as a future possibility, out of scope of this spec.